packages/ws/src/utils/WorkerBootstrapper.ts

```diff
@@ -0,0 +1,160 @@
+	/**
+	 * Function to call when a shard is created for additional setup
+	 */
+	shardCallback?(shard: WebSocketShard): Awaitable<void>;
+}
+
+/**
+ * Utility class for bootstraping a worker thread to be used for sharding
+ */
+export class WorkerBootstrapper {
+	/**
+	 * The data passed to the worker thread
+	 */
+	protected readonly data = workerData as WorkerData;
+
+	/**
+	 * The shards that are managed by this worker
+	 */
+	protected readonly shards = new Collection<number, WebSocketShard>();
+
+	public constructor() {
+		if (isMainThread) {
+			throw new Error('Expected WorkerBootstrap to not be used within the main thread');
+		}
+	}
+
+	/**
+	 * Helper method to initiate a shard's connection process
+	 */
+	protected async connect(shardId: number): Promise<void> {
+		const shard = this.shards.get(shardId);
+		if (!shard) {
+			throw new Error(`Shard ${shardId} does not exist`);
+		}
+
+		await shard.connect();
+	}
+
+	/**
+	 * Helper method to destroy a shard
+	 */
+	protected async destroy(shardId: number, options?: WebSocketShardDestroyOptions): Promise<void> {
+		const shard = this.shards.get(shardId);
+		if (!shard) {
+			throw new Error(`Shard ${shardId} does not exist`);
+		}
+
+		await shard.destroy(options);
+	}
+
+	/**
+	 * Helper method to attach event listeners to the parentPort
+	 */
+	protected setupThreadEvents(): void {
+		parentPort!
+			.on('messageerror', (err) => {
+				throw err;
+			})
+			.on('message', async (payload: WorkerSendPayload) => {
+				switch (payload.op) {
+					case WorkerSendPayloadOp.Connect: {
+						await this.connect(payload.shardId);
+						const response: WorkerRecievePayload = {
+							op: WorkerRecievePayloadOp.Connected,
+							shardId: payload.shardId,
+						};
+						parentPort!.postMessage(response);
+						break;
+					}
+
+					case WorkerSendPayloadOp.Destroy: {
+						await this.destroy(payload.shardId, payload.options);
+						const response: WorkerRecievePayload = {
+							op: WorkerRecievePayloadOp.Destroyed,
+							shardId: payload.shardId,
+						};
+
+						parentPort!.postMessage(response);
+						break;
+					}
+
+					case WorkerSendPayloadOp.Send: {
+						const shard = this.shards.get(payload.shardId);
+						if (!shard) {
+							throw new Error(`Shard ${payload.shardId} does not exist`);
+						}
+
+						await shard.send(payload.payload);
+						break;
+					}
+
+					case WorkerSendPayloadOp.SessionInfoResponse: {
+						break;
+					}
+
+					case WorkerSendPayloadOp.ShardCanIdentify: {
+						break;
+					}
+				}
+			});
+	}
+
+	/**
+	 * Bootstraps the worker thread with the provided options
+	 */
+	public async bootstrap(options: BootstrapOptions = {}): Promise<void> {
+		// Start by initializing the shards
+		for (const shardId of this.data.shardIds) {
+			const shard = new WebSocketShard(new WorkerContextFetchingStrategy(this.data), shardId);
+			for (const event of (options.forwardEvents ??= Object.values(WebSocketShardEvents))) {
```

Comment:
You're still mutating. `??=` → `??`.

To be safe, we should do `Readonly<BootstrapOptions>`, akin to what we do in `/core`.